MINOR: Allow single struct to be a field in protocol spec

soondenana opened the pull request:
Currently protocol generation code has this restriction that a structure
field need to be an array. There are however some case where fields need
to be grouped but not be an array. This is for organizational purposes
and mirrors program structure.

This change adds this capability to the protocol generation.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
MINOR: Allow single struct to be a field in protocol spec

Currently protocol generation code has this restriction that a structure
field need to be an array. There are however some case where fields need
to be grouped but not be an array. This is for organizational purposes
and mirrors program structure.

This change adds this capability to the protocol generation.

cmccabe commented on generator/src/main/java/org/apache/kafka/message/FieldSpec.java:
```diff
@@ -101,11 +101,7 @@ public FieldSpec(@JsonProperty("name") String name,
-                throw new RuntimeException("Non-array field " + name + " cannot have fields");
-            }
-        }
+
```
I don't think we should get rid of this check entirely.  We should change it to check for either array or struct type, rather than just array type, right?

It wouldn't make sense to have an int32 with fields, for example.

soondenana replied:
Makes sense. Done.

cmccabe commented:
Thanks for tackling this!  The change looks good as far as it goes.  I left a comment.

You'll want to add an example of a struct to SimpleExampleMessage.json and modify SimpleExampleMessageTest to make sure that round-trip serialization and deserialization works.

soondenana committed:
Take care of review comments

soondenana commented:
> Thanks for tackling this! The change looks good as far as it goes. I left a comment.
> 
> You'll want to add an example of a struct to SimpleExampleMessage.json and modify SimpleExampleMessageTest to make sure that round-trip serialization and deserialization works.

Thanks Colin for the pointer. Added more tests, and it exposed equality check that wasn't changed. Fixed that too. Please take a look at latest commit.

soondenana committed:
Remove Struct nullable

Making struct nullable is a bigger change, my previous attempt left a
bunch of place where modificaiton is needed. Simpler to not support it
as this is the current state.

cmccabe commented:
ok to test

cmccabe commented:
Looks good, @soondenana!  Just one more thing... can you add a test of a tagged struct in  MessageTest.java?  I think using tagged structs was one of the use-cases we really wanted to support.

LGTM after that is done

soondenana committed:
Add support for tagged structs

soondenana commented:
> Looks good, @soondenana! Just one more thing... can you add a test of a tagged struct in MessageTest.java? I think using tagged structs was one of the use-cases we really wanted to support.

Done. Needed to add additional if conditions to part where we generate tagged field.

soondenana committed:
Add roundtrip test for SimpleExampleMessage

cmccabe commented:
ok to test

cmccabe commented:
retest this please

cmccabe commented:
retest this please

soondenana commented:
So there is one test failure for JDK8 (JDK 11 passes):

> kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup

with error:
> org.scalatest.exceptions.TestFailedException: The remaining consumers in the group could not fetch the expected records

As this doesn't relate to this change, I think this should be good to merge.

cmccabe commented:
LGTM
